### PR TITLE
EE-8289: Checkpoint for design review

### PR DESCRIPTION
**Pull request by Tim-Hammonds**

### CODE REVIEW

File: src/test/resources/specs/name-matching/analysis/Last_names_combination_meta_data.feature [hunk 1]
```diff
@@ -0,0 +1,390 @@
...
+    And the meta-data indicates that the following generators were used
+      | SPLITTERS_REPLACED           |
+      | NAME_MATCHING                |
+    And the meta-data contains the following name derivation information
+      | section   | index   | length  | derivationActions |
+      | FIRST     | 0       | 6       | ORIGINAL          |
+      | LAST      | 0       | 7       | RIGHT_OF_SPLIT    |
+
+
+  Scenario: Meta-Data for Apostrophed last name is tried without apostrophes 4
```
Comment (mikehalmaho) [inline src/test/resources/specs/name-matching/analysis/Last_names_combination_meta_data.feature]: I know this is not related to the meta-data collection, but I wonder if the match for this test is correct - I don't think candidate name "Arthur O'Bobbins" would get a match with HMRC name "Arthur O".

On a plus note, the meta-data makes it much easier to understand edge cases like this :)

File: src/main/java/uk/gov/digital/ho/pttg/application/HmrcHateoasClient.java [hunk 2]
```diff
@@ -218,7 +218,13 @@ private String stripPlaceholderQueryParams(String href) {
...
             try {
                 final Resource<String> matchedIndividual = performMatchedIndividualRequest(matchUrl, accessToken, candidateNames.get(retries), individual.getNino(), individual.getDateOfBirth());
-                log.info("Successfully matched individual {}", individual.getNino(), value(EVENT, HMRC_MATCHING_SUCCESS_RECEIVED));
+
+                log.info("Successfully matched individual {}",
+                        individual.getNino(),
+                        value("combination", String.format("%d of %d", retries + 1, candidateNames.size())),
+                        value("name-matching-analysis", candidateNames.get(retries).derivation()),
+                        value(EVENT, HMRC_MATCHING_SUCCESS_RECEIVED));
+
```
Comment (mikehalmaho) [inline src/main/java/uk/gov/digital/ho/pttg/application/HmrcHateoasClient.java]: Is there some way we can log out the input names analysis in the case that we never get a match?  Probably requires a new ticket...

File: src/test/java/bdd/steps/NameMatchingSteps.java [hunk 3]
```diff
@@ -555,4 +582,162 @@ private static ObjectMapper setupObjectMapper() {
...
+        CandidateDerivation candidateDerivation = getCandidateDerivation(loggingEvent);
+        return generators.equals(candidateDerivation.generators());
+    }
+
+    private CandidateDerivation getCandidateDerivation(LoggingEvent loggingEvent) {
+        return (CandidateDerivation) ReflectionTestUtils.getField(loggingEvent.getArgumentArray()[2], "object");
+    }
+
+    @And("^the meta-data contains the following name derivation information$")
+    public void theMetaDataContainsTheFollowingNameDerivationInformation(DataTable dataTable) {
```
Comment (mikehalmaho) [inline src/test/java/bdd/steps/NameMatchingSteps.java]: The diagnosis done on the input names and generators makes test failures easier to understand - can we do the same with the name derivations?

File: src/main/java/uk/gov/digital/ho/pttg/application/HmrcHateoasClient.java (hunk 2 above)
Comment (Tim-Hammonds) [reply src/main/java/uk/gov/digital/ho/pttg/application/HmrcHateoasClient.java]: I've created this JIRA: https://jira.ipttools.info/browse/EE-11594

File: src/test/resources/specs/name-matching/analysis/Last_names_combination_meta_data.feature (hunk 1 above)
Comment (Tim-Hammonds) [reply src/test/resources/specs/name-matching/analysis/Last_names_combination_meta_data.feature]: I'll chase this up with Owain.

File: src/main/java/uk/gov/digital/ho/pttg/application/namematching/candidates/NameMatchingCandidateGenerator.java [hunk 4]
```diff
@@ -7,14 +7,17 @@
...
-    int ALIAS_COMBINATIONS_STRATEGY_PRIORITY = 5;
-    int NAME_MATCHING_STRATEGY_PRIORITY = 6;
-    int SPECIAL_CHARACTERS_STRATEGY_PRIORITY = 7;
+    enum Generator {
+        ENTIRE_NON_ALIAS_NAME,
+        ENTIRE_LAST_NAME_AND_EACH_FIRST_NAME,
+        MULTIPLE_NAMES,
+        ABBREVIATED_NAMES,
+        ALIAS_COMBINATIONS,
+        NAME_MATCHING,
```
[the remark this answers is not in the document]
Comment (Tim-Hammonds) [reply src/main/java/uk/gov/digital/ho/pttg/application/namematching/candidates/NameMatchingCandidateGenerator.java]: Agreed

File: src/test/java/bdd/steps/NameMatchingSteps.java (hunk 3 above)
Comment (Tim-Hammonds) [reply src/test/java/bdd/steps/NameMatchingSteps.java]: Agreed

File: src/test/resources/specs/name-matching/analysis/Last_names_combination_meta_data.feature [hunk 5]
```diff
@@ -0,0 +1,390 @@
...
+      | nameType  | index | diacritics  | umlauts | abbreviation  | nameSplitter  | uniCodeBlocks | length  |
+      | FIRST     | 0     | false       | false   | false         | true          | BASIC_LATIN   | 12      |
+      | LAST      | 0     | false       | false   | false         | true          | BASIC_LATIN   | 8       |
+    And the meta-data indicates that the following generators were used
+      | SPLITTERS_REPLACED  |
+      | NAME_MATCHING       |
+    And the meta-data contains the following name derivation information
+      | section   | index   | length  | derivationActions |
+      | FIRST     | 0       | 6       | LEFT_OF_SPLIT     |
+      | LAST      | 0       | 6       | RIGHT_OF_SPLIT    |
```
[the remark this answers is not in the document]
Comment (Tim-Hammonds) [reply src/test/resources/specs/name-matching/analysis/Last_names_combination_meta_data.feature]: Bug detected - now fixed.

File: src/test/resources/specs/name-matching/analysis/Last_names_combination_meta_data.feature [hunk 6]
```diff
@@ -0,0 +1,390 @@
...
+    And the meta-data indicates that the following generators were used
+      | SPLITTERS_REPLACED  |
+      | NAME_MATCHING       |
+    And the meta-data contains the following name derivation information
+      | section   | index   | length  | derivationActions |
+      | FIRST     | 0       | 5       | RIGHT_OF_SPLIT    |
+      | LAST      | 0       | 6       | RIGHT_OF_SPLIT    |
+
+
+  Scenario: Meta-Data for Apostrophed and Hyphenated names are tried without apostrophes and hyphens 6
```
[the remark this answers is not in the document]
Comment (Tim-Hammonds) [reply src/test/resources/specs/name-matching/analysis/Last_names_combination_meta_data.feature]: I think so. The significant characters match with the HMRC record. The data after that is ignored. I think? One to check with Owain.

File: src/test/resources/specs/name-matching/analysis/Alias_name_matching_meta_data.feature [hunk 7]
```diff
@@ -0,0 +1,172 @@
+@jira=EE-8289
+
+Feature: Analysis logging
+
```
[the remark this answers is not in the document]
Comment (Tim-Hammonds) [reply src/test/resources/specs/name-matching/analysis/Alias_name_matching_meta_data.feature]: Tests added
